unit-tests/log/test-multi-thread-logs.cpp

```diff
@@ -0,0 +1,231 @@
+const int checked_lag_time = 15; //ms
+const int required_lag_time = 50; //ms
+//----------------------- HELPER METHODS START -------------------------
+int get_time_from_line(std::string line)
+{
+    auto time_word = line.substr(7, 12);
+    auto hours = stoi(time_word.substr(0, 2));
+    auto minutes = stoi(time_word.substr(3, 5));
+    auto seconds = stoi(time_word.substr(6, 8));
+    auto ms = stoi(time_word.substr(9, 12));
+    //std::cout << "h:m:s:ms = " << hours << ":" << minutes << ":" << seconds << ":" << ms << std::endl;
+    return (ms + seconds * 1000 + minutes * 1000 * 60 + hours * 1000 * 60 * 60);
+}
+
+
+std::string get_last_line(std::ifstream& in)
+{
+    std::string line;
+    for (int i = 0; i < (number_of_iterations * number_of_threads) - 1; ++i)
+    {
+        getline(in, line);
+    }
+    return line;
+}
+
+// inputs are use as 2 points that define a line:
+// first_time_ms, first_value : x1, y1
+// second_time_ms, second_value : x2, y2
+// value_in_log : y3
+// return value will be x3
+int get_log_ideal_time(int first_time_ms, int first_value, int second_time_ms, int second_value, int value_in_log)
+{
+    int ideal_time = value_in_log * (second_time_ms - first_time_ms) / (second_value - first_value) +
+        (first_time_ms * (second_value / (second_value - first_value))) - second_time_ms * (first_value / (second_value - first_value));
+    
+    return ideal_time;
+}
+
+std::pair<unsigned int, int> get_next_50_lines_time_val(std::ifstream& ifs)
+{
+    unsigned int avg_time_ms = -1;
+    int avg_value = -1;
+    for (int i = 0; i < 50; ++i) {
+        std::string line;
+        getline(ifs, line);
+        avg_time_ms += get_time_from_line(line);
+        avg_value += stoi(line.substr(line.find_last_of(" ")));
+    }
+    avg_time_ms /= 50;
+    avg_value /= 50;
+    return std::make_pair(avg_time_ms, avg_value);
+}
+
+//----------------------- HELPER METHODS END ---------------------------
+
+TEST_CASE("async logger", "[log][async_log]")
+{
+    size_t n_callbacks = 0;
+    auto callback = [&](rs2_log_severity severity, rs2::log_message const& msg)
+    {
+        ++n_callbacks;
+        std::cout << msg.raw();
+    };
+    
+    auto func = [](int required_value) {
+        int iterations = 0;
+        auto start = std::chrono::steady_clock::now();
+        
+        while (iterations < number_of_iterations)
+        {
+            std::stringstream ss;
+            ss << "atomic integer = " << ++atomic_integer;
+            auto start_ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+            rs2::log(RS2_LOG_SEVERITY_DEBUG, ss.str().c_str());
+            std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+            std::chrono::milliseconds delta_ms = ms - start_ms;
+            if (delta_ms > max_time) max_time = delta_ms;
```

Comment:
Shouldn't these be atomic?

Reply:
Well, it seems you are right, but since these are chrono and not built-in types, I get some errors when I use the count() api if I make them atomic. In the worst case, I might miss the minimum or the maximum.
The other option is to use a mutex there, but it might get the threads slower, which I prefer to avoid.

Reply:
But you need some form of sync to ensure correct results, no? Otherwise the unit-test may sometimes fail...

Reply:
How can a race condition can make this test fail? Let's talk about it.

Reply:
lock_guard added